posthog/temporal/README.md

```diff
@@ -1,20 +1,303 @@
+
+#### Multithreading
+
+There shouldn't be ever a need to run synchronous multithreading activities, as you can always start a thread pool in asynchronous activities when required. Regardless, if you do have the need, then know that just making an activity function synchronous is enough as our workers are configured with a thread pool executor by default. Similar considerations around blocking as in asyncio exist here too: The GIL will prevent other activities from running unless released, so your threaded code should contain I/O operations that release the GIL for this concurrency model to make sense.
+
+Using synchronous activities means that you have to opt for synchronous versions of all of our abstractions, like `HeartbeaterSync` instead of `Heartbeater`.
+
+#### Multiprocessing
+
+
+
+### Timing-out, heart-beating, and retrying activities
+
+Temporal allows us to apply multiple timeouts to activities:
+1. Schedule-to-close: Time out based on the time from the moment the Temporal service puts an activity task in its queue.
+2. Start-to-close: Time out based on the time from the moment a worker starts executing an activity.
+3. Schedule-to-start: Time out based on the time it takes for an activity to be picked up by a worker.
+4. Heartbeat: Time out based on time between heartbeats.
+
+Every activity **must** have at least one of schedule-to-close and start-to-close timeouts defined, and I recommend the latter. This is how the Temporal service recovers after worker crashes: After the timeout expires, the service will re-issue the activity so that it can be picked up by surviving workers.
+
+However, for long-running activities, these two timeouts are not enough: Imagine an activity that we expect to take 1 hour to complete, so we set a `start_to_close` timeout of 1 hour. If the worker crashes as soon as the activity begins, we will need to wait almost the full hour until the time out expires and the service re-schedules it. This is too long: We essentially wasted that whole hour, and potentially are now backed up as the new hour begins and we have a new workflow starting. This is why heartbeating and heartbeat timeouts are strongly recommended for any long running activities. By emitting heartbeats, the activities let the service know that they are still alive, and when they stop heartbeating for the duration of the timeout the service knows that the worker has likely crashed and the activity can be retried immediately, without waiting the full start-to-close timeout.
+
+Implementing hearbeating is very easy with the help of the `posthog.temporal.common.heartbeat.Heartbeater` class. This can be used as a context manager to wrap your long running work. `Heartbeater` will schedule a task that issues a heartbeat to the Temporal service within your configured `heartbeat_timeout`.
```

Comment:
Not sure if it's worth adding a note about heartbeat details, i.e. that you can add certain data to heartbeats to enable recovery in the case of retries? Or maybe this belongs in a section about worker shutdown and retries?

Reply:
Yeah I agree. Maybe we could add a note to say it's possible to include information such as tracking progress, but after using it in practice we don't recommend it, at least for anything critical, as it's too unreliable

Reply:
I think it would also be worth including some information about the shutdown monitor since that's a pretty cool feature you added